docs/2.10.0/docs/canton/usermanual/kms/kms_driver_guide.rst

```diff
@@ -0,0 +1,270 @@
+
+``src/main/resources/META-INF/services/com.digitalasset.canton.crypto.kms.driver.api.v1.KmsDriverFactory``
+
+and in the following file for the base driver factory (without v1):
+
+``src/main/resources/META-INF/services/com.digitalasset.canton.crypto.kms.driver.api.KmsDriverFactory``
+
+The major part of the implementation is the ``v1.KmsDriver`` that is specific to
+the KMS/HSM to be integrated with. The supported key and algorithm
+specifications can be defined statically depending on the capabilities of the
+underlying KMS/HSM. To ensure the best compatibility with other Canton nodes,
+all currently specified key and algorithm specifications should be supported.
+
+Any credentials that are required by the underlying KMS/HSM can either be passed
+through the Canton configuration file as part of the driver-specific
+configuration, where secrets can be resolved from the environment, or the driver
+can retrieve credentials directly from the environment or any other
+driver-specific means.
+
+Bundle your driver into a self-contained jar, that is, with all required
+libraries included in the jar. That way you only need a single driver jar when
+starting Canton with your KMS Driver.
+
+KMS Driver Testing
+~~~~~~~~~~~~~~~~~~
+
+The reusable test suite for KMS Drivers is published at
+`canton-kms-driver-testing
+<https://digitalasset.jfrog.io/ui/repos/tree/General/canton-kms-driver-testing>`__.
+Configure your build system to depend on this maven artifact in the test scope
+of your project (e.g. for sbt append % Test to limit the dependency to the test
+scope).
+
+KmsDriverTest
+^^^^^^^^^^^^^
+
+The main part of the test suite is the ``KmsDriverTest`` that tests the
+functionality of a driver against the ``KmsDriver`` API.
+
+In the simplest form the specific driver test class extends the
+``KmsDriverTest`` and allows the generation of new keys as part of the test:
+
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/aws-kms-driver/src/test/scala/com/digitalasset/canton/nightly/AwsKmsDriverTest.scala
+   :language: Scala
+   :start-after: user-manual-entry-begin: AwsKmsDriverTest
+   :end-before: user-manual-entry-end: AwsKmsDriverTest
+
+Given that generation of new keys can be expensive when running tests during
+development frequently, in particular when using Cloud-based KMSs, the test
+suite can also be configured to use predefined keys to test most parts (except
+key generation obviously) of the KMS Driver API:
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/aws-kms-driver/src/test/scala/com/digitalasset/canton/nightly/AwsKmsDriverTest.scala
+   :language: Scala
+   :start-after: user-manual-entry-begin: AwsKmsDriverWithPredefinedKeysTest
+   :end-before: user-manual-entry-end: AwsKmsDriverWithPredefinedKeysTest
+
+
+For each supported signing/encryption key specification an existing key alias/id
+can be configured as part of the predefined keys maps. When running the test
+suite the generation of new keys is not allowed.
```

Comment:
```suggestion
For each supported signing/encryption key specification an existing key alias/ID
can be configured as part of the predefined key maps. When running the test
suite the generation of new keys is not allowed.
```

Reply:
there can be multiple keys in each map, so I kept the plural and only upper-cased the ID.